### ISSUE TITLE
Duplicate Content-Length header not handled correctly

### ISSUE THREAD

**Issue by sbordet** [eclipse/jetty.project#423]
The relevant section of the specification is: https://tools.ietf.org/html/rfc7230#section-3.3.2.

```
POST / HTTP/1.1
Host: localhost
Content-Length: 1
Content-Length: 2
```

The request should be rejected.

**Comment by sbordet** [comment]
Fixed as required by the spec, rejecting the request.

**Comment by joakime** [comment]
What

**Comment by joakime** [comment]
This should probably be configurable via a new HttpCompliance option on the parser.

**Comment by joakime** [comment]
@djscholl frankly, I'm surprised that those problematic clients haven't been impacted by all of the various intermediaries that also reject that kind of request.

**Comment by djscholl** [comment]
Thanks for the response @joakime. It's likely that these problematic clients do run into other problems. We found at least one existing problematic test client in our testing of the upgrade. The main concern for us is the unknown factor of how often these sorts of requests end up getting all the way to our server. We could add some logging to start detecting how often that happens, but we don't have the historical data to give us any confidence that it never happens.

**Comment by djscholl** [comment]
@joakime or anyone else - I assume this was done as part of a general cleanup of "invalid" requests. Is there any documented list of all the related changes that were made? We also found that GET requests with non-zero content lengths had their handling changed. The new handling makes sense (Jetty waits for the data and the request eventually times out if there is none). But since we've found two new behaviors, we'd like to try and understand any others that are yet to be discovered. I appreciate any help anyone can give! We're hoping to not have to crawl all of the git commits to find these answers.

**Comment by joakime** [comment]
Bad HTTP clients that rely on bad spec interpretation or rely quirks in server behavior are always subject to being broken when the various intermediaries and servers tighten up their behaviors.
If the HTTP client follows the specs, then breakage is truly rare beast.  (And if discovered cause a quick fix on our side!)

As for a `GET` request with a payload, this is problematic, as this concept no spec defined behavior.

The [RFC7231: Section 4.3.1](https://tools.ietf.org/html/rfc7231#section-4.3.1) states:

```
   A payload within a GET request message has no defined semantics;
   sending a payload body on a GET request might cause some existing
   implementations to reject the request.
```

Jetty has traditionally supported GET requests with payload, with minor differences on how HttpServletRequest behaves based on the method (GET vs POST/PUT).

It's worth noting that many intermediaries can also make a judgement call on allowing or rejecting GET with request payloads as well (firewalls, load balancers, mobile network http caches, mobile OS http optimization, etc).

As for the list of changes, the VERSION.txt is the list of changes in high level form.
The individual issues / commits / test cases document the behavior listed in the VERSION.txt for each release.

Just to repeat, we have put many of these more problematic changes behind various `HttpCompliance` configurables in the existing codebase.

See ...

* [org/eclipse/jetty/http/HttpCompliance](https://github.com/eclipse/jetty.project/blob/jetty-9.4.11.v20180605/jetty-http/src/main/java/org/eclipse/jetty/http/HttpCompliance.java) - for high level grouping of various compliance behaviors
* [org/eclipse/jetty/http/HttpComplianceSection](https://github.com/eclipse/jetty.project/blob/jetty-9.4.11.v20180605/jetty-http/src/main/java/org/eclipse/jetty/http/HttpComplianceSection.java) - for the specific behaviors in parsing / generating that have a switch in behavior.
* [org/eclipse/jetty/http/CookieCompliance](https://github.com/eclipse/jetty.project/blob/jetty-9.4.11.v20180605/jetty-http/src/main/java/org/eclipse/jetty/http/CookieCompliance.java) - even Cookie parsing/generating behavior has been modified over time to behave to the newer, stricter, Cookie standards that nearly all modern browsers now require.

**Comment by cy6erGn0m** [comment]
Isn't it related to https://security-tracker.debian.org/tracker/CVE-2017-7657 ?

**Comment by joakime** [comment]
@cy6erGn0m that CVE was fixed with commit a285deea42fcab60d9edcf994e458c238a348b55 (Issue #2529)
See June 2018 announcement for all recent CVEs (including the one you linked): https://dev.eclipse.org/mhonarc/lists/jetty-announce/msg00123.html

**Comment by joakime** [comment]
This is currently implemented in branch `jetty-9.4.x` and has been in place since commit a285deea42fcab60d9edcf994e458c238a348b55
Available in Jetty 9.4.11.v20180605 and newer.
